File: dart/test/unicode_rule_label_identity_routes_test.dart
// FUTURE-PARITY-BACKLOG.10.5.0.2.2 — exact downstream Dart label identity.

import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_dart/src/cli/primary_cli.dart';
import 'package:test/test.dart';

final Map<String, Object?> _contract =
    jsonDecode(
          File(
            '../capability_conformance/unicode_rule_label_contract.json',
          ).readAsStringSync(),
        )
        as Map<String, Object?>;

final List<Map<String, Object?>> _positiveFixtures = [
  for (final fixture
      in (_contract['positive_fixtures']! as List).cast<Map<String, Object?>>())
    Map<String, Object?>.from(fixture),
];

final List<String> _positiveLabels = [
  for (final fixture in _positiveFixtures) fixture['label']! as String,
];

final List<Map<String, Object?>> _distinctFixtures = [
  for (final fixture
      in (_contract['distinct_fixtures']! as List).cast<Map<String, Object?>>())
    Map<String, Object?>.from(fixture),
];

final List<String> _allLabels = <String>{
  ..._positiveLabels,
  for (final fixture in _distinctFixtures) fixture['left']! as String,
  for (final fixture in _distinctFixtures) fixture['right']! as String,
}.toList(growable: false);

final String _source = _sourceForLabels(_allLabels);

void main() {
  test('every positive and distinct label survives compiled artifacts', () {
    expect(_positiveLabels, hasLength(9));
    expect(_distinctFixtures, hasLength(2));

    final parsed = parseSpec(_source);
    validateSpec(parsed);
    final compiled = compileSpec(parsed);
    final compiledJson = compiled.toJson();
    final descriptor = compiled.toDescriptorJson();
    final plan = buildGeneratedRulePlan(compiled);

    expect(parsed.rules.map((rule) => rule.header.label), _allLabels);
    expect(compiled.definitionOrder, _allLabels);
    expect(compiled.compiledRuleOrder, _allLabels);
    expect(compiled.rulesByLabel.keys, _allLabels);
    expect(
      (compiledJson['rules_by_label']! as Map<String, Object?>).keys,
      _allLabels,
    );
    expect((descriptor['spec']! as Map<String, Object?>).keys, _allLabels);
    final descriptorMeta = descriptor['meta']! as Map<String, Object?>;
    expect(descriptorMeta['definition_order'], _allLabels);
    expect(descriptorMeta['compiled_rule_order'], _allLabels);
    expect(plan.map((row) => row.label), _allLabels);

    final reconstructed = compileSpec(
      SpecFile.fromJson(
        Map<String, Object?>.from(
          jsonDecode(jsonEncode(parsed.toJson())) as Map,
        ),
      ),
    );
    expect(reconstructed.toJson(), compiledJson);
    expect(reconstructed.toDescriptorJson(), descriptor);

    for (final label in _allLabels) {
      expect(compiled.rule(label)?.label, label);
      expect(
        LinkedSpecRuntimeEngine(
          reconstructed,
        ).execute('x', topRule: label).value,
        label,
      );
      expect(
        executeGeneratedParserV2(
          compiled,
          plan,
          'x',
          'unicode-label/direct-plan.spec',
          topRule: label,
        ),
        label,
      );
    }

    for (final fixture in _distinctFixtures) {
      final left = fixture['left']! as String;
      final right = fixture['right']! as String;
      expect(left, isNot(right), reason: fixture['id']! as String);
      expect(compiled.rule(left), isNotNull);
      expect(compiled.rule(right), isNotNull);
      expect(compiled.rule(left), isNot(same(compiled.rule(right))));
    }
  });

  test(
    'emitted source reconstructs and executes every exact label',
    () async {
      const identity = 'unicode-label/emitted-規則-𐐀.spec';
      final compiled = compileSpec(parseSpec(_source));
      final emitted = emitDartSourceV2(compiled, identity);
      final encodedPayload = RegExp(
        "const _compiledSpecJsonBase64 = '([^']+)';",
      ).firstMatch(emitted)![1]!;
      final payload =
          jsonDecode(
                utf8.decode(
                  base64Decode(encodedPayload),
                  allowMalformed: false,
                ),
              )
              as Map<String, Object?>;
      final reconstructed = compileSpec(SpecFile.fromJson(payload));
      expect(reconstructed.compiledRuleOrder, _allLabels);
      expect(
        buildGeneratedRulePlan(reconstructed).map((row) => row.label),
        _allLabels,
      );

      final scratch = Directory.systemTemp.createTempSync(
        'linkedspec-dart-unicode-label-emitted-',
      );
      final packageRoot = Directory.current.absolute;
      final pubCache = Directory('${scratch.path}/pub-cache')..createSync();
      addTearDown(() {
        if (scratch.existsSync()) {
          scratch.deleteSync(recursive: true);
        }
      });

      Directory('${scratch.path}/lib').createSync();
      Directory('${scratch.path}/bin').createSync();
      File('${scratch.path}/pubspec.yaml').writeAsStringSync('''
name: linkedspec_unicode_label_emitted_probe
publish_to: none
environment:
  sdk: ">=3.9.0 <4.0.0"
dependencies:
  linkedspec_dart:
    path: ${jsonEncode(packageRoot.path)}
''');
      File('${scratch.path}/lib/generated.dart').writeAsStringSync(emitted);
      File('${scratch.path}/bin/main.dart').writeAsStringSync('''
import 'dart:convert';

import 'package:linkedspec_unicode_label_emitted_probe/generated.dart'
    as generated;

void main() {
  final labels = <String>[${_allLabels.map(jsonEncode).join(', ')}];
  final plan = generated.plan();
  generated.validatePlan(plan);
  final values = <Object?>[
    for (final label in labels) generated.execute('x', topRule: label),
  ];
  print(jsonEncode({
    'identity': generated.metadata().sourceIdentity,
    'labels': labels,
    'plan': [for (final row in plan) row.label],
    'values': values,
  }));
}
''');

      final environment = {...Platform.environment, 'PUB_CACHE': pubCache.path};
      await _expectProcessSuccess(scratch, environment, const [
        'pub',
        'get',
        '--offline',
      ]);
      await _expectProcessSuccess(scratch, environment, const [
        'analyze',
        '--fatal-infos',
        '--fatal-warnings',
      ]);
      final run = await _expectProcessSuccess(scratch, environment, const [
        'run',
        'bin/main.dart',
      ]);
      expect(jsonDecode((run.stdout as String).trim()), {
        'identity': identity,
        'labels': _allLabels,
        'plan': _allLabels,
        'values': _allLabels,
      });
    },
    timeout: const Timeout(Duration(minutes: 2)),
  );

  test('strict loading and primary commands preserve every exact label', () {
    final scratch = Directory.systemTemp.createTempSync(
      'linkedspec-dart-unicode-label-primary-',
    );
    addTearDown(() {
      if (scratch.existsSync()) {
        scratch.deleteSync(recursive: true);
      }
    });
    final specFile = File('${scratch.path}/unicode.spec')
      ..writeAsBytesSync(utf8.encode(_source));
    final loaded = loadAndCompileSpec(
      SpecRequest.path(specFile.path),
      SpecLoadOptions(cwd: scratch),
    );

    expect(loaded.loaded.sourceText, _source);
    expect(loaded.compiled.compiledRuleOrder, _allLabels);
    for (final label in _allLabels) {
      expect(loaded.createEngine().execute('x', topRule: label).value, label);
      final inline = runLinkedSpecDartPrimaryCli([
        '--inline-spec',
        _source,
        '--input',
        'x',
        '--top-rule',
        label,
      ]);
      expect(inline.exitCode, 0, reason: label);
      expect(inline.stderrBytes, isEmpty, reason: label);
      expect(
        inline.stdoutBytes,
        utf8.encode('${jsonEncode(label)}\n'),
        reason: label,
      );
    }

    final supplementary =
        _positiveFixtures.singleWhere(
              (fixture) => fixture['id'] == 'supplementary',
            )['label']!
            as String;
    final fileRoute = runLinkedSpecDartPrimaryCli(
      [
        '--spec-file',
        specFile.path,
        '--input',
        'x',
        '--top-rule',
        supplementary,
      ],
      workingDirectory: scratch,
      repositoryRoot: Directory('..'),
    );
    expect(fileRoute.exitCode, 0);
    expect(fileRoute.stderrBytes, isEmpty);
    expect(
      fileRoute.stdoutBytes,
      utf8.encode('${jsonEncode(supplementary)}\n'),
    );
  });

  test('selectors diagnostics and traces retain exact Unicode identity', () {
    final selected =
        _distinctFixtures.singleWhere(
              (fixture) => fixture['id'] == 'normalization_sensitive',
            )['right']!
            as String;
    const missing = '規則Missing𐐀';
    const identity = 'unicode-label/diagnostic-規則.spec';
    final compiled = compileSpec(parseSpec(_source));
    final engine = LinkedSpecRuntimeEngine(compiled);
    final plan = buildGeneratedRulePlan(compiled);
    final stdout = StringBuffer();
    final trace = LinkedSpecTraceEmitter(
      LinkedSpecTraceConfig.enabled(LinkedSpecTraceLevel.debug),
      stdoutWriter: stdout.write,
    );

    expect(
      engine.execute('x', topRule: selected, trace: trace).value,
      selected,
    );
    expect(
      trace.events.any(
        (event) =>
            event.topic == 'dart_runtime:entry_rule_selection' &&
            event.details.contains('requested=$selected') &&
            event.details.contains('effective=$selected') &&
            event.details.contains('basis=explicit_selector'),
      ),
      isTrue,
    );
    expect(trace.lines.join(), contains('top_rule=$selected'));
    expect(stdout.toString(), contains(selected));

    final nativeFailure = _runtimeFailure(
      () => engine.execute('x', topRule: missing, trace: trace),
    );
    expect(nativeFailure.diagnostic?.toJson(), {
      'type': 'runtime_parser',
      'stage': 'select_entry_rule',
      'owner_stage': 'dart_runtime',
      'summary': 'Dart runtime entry-rule selection failed',
      'detail': "entry rule '$missing' is not defined",
      'code': 'entry_rule_not_found',
      'top_rule': missing,
      'entry_rule': missing,
      'rule_label': missing,
      'handler_source_label': 'dart_runtime:rule:$missing',
    });
    expect(trace.lines.join(), contains('requested=$missing'));

    expect(
      () => executeGeneratedParserV2(
        compiled,
        plan,
        'x',
        identity,
        topRule: missing,
      ),
      throwsA(
        isA<GeneratedSourceException>().having(
          (error) => error.toJson(),
          'portable generated selector diagnostic',
          {
            'type': 'generated_source_error',
            'stage': 'select_entry_rule',
            'code': 'entry_rule_not_found',
            'summary': 'Generated Dart parser entry-rule selection failed',
            'source_identity': identity,
            'entry_rule': missing,
            'rule_label': missing,
            'detail': "entry rule '$missing' is not defined",
          },
        ),
      ),
    );

    final scratch = Directory.systemTemp.createTempSync(
      'linkedspec-dart-unicode-label-trace-',
    );
    addTearDown(() {
      if (scratch.existsSync()) {
        scratch.deleteSync(recursive: true);
      }
    });
    final traceFile = File('${scratch.path}/generated.trace');
    expect(
      executeGeneratedParserWithTraceV2(
        compiled,
        plan,
        'x',
        LinkedSpecTraceConfig(
          level: LinkedSpecTraceLevel.debug,
          traceFile: traceFile.path,
          sinkMode: LinkedSpecTraceSinkMode.route,
          resetFile: true,
        ),
        identity,
        topRule: selected,
      ),
      selected,
    );
    final generatedTrace = traceFile.readAsStringSync();
    expect(generatedTrace, contains('requested=$selected'));
    expect(generatedTrace, contains('effective=$selected'));
    expect(generatedTrace, contains('top_rule=$selected'));
    expect(generatedTrace, contains(identity));
  });
}

String _sourceForLabels(List<String> labels) {
  final source = StringBuffer();
  for (var index = 0; index < labels.length; index += 1) {
    final label = labels[index];
    source
      ..writeln('$label${index == 0 ? '::' : ':'}')
      ..writeln(' /x/')
      ..writeln(' E { return(${jsonEncode(label)}) }')
      ..writeln();
  }
  return source.toString();
}

RuntimeInterpreterException _runtimeFailure(Object? Function() operation) {
  try {
    operation();
  } on RuntimeInterpreterException catch (error) {
    return error;
  }
  throw StateError('expected runtime failure');
}

Future<ProcessResult> _expectProcessSuccess(
  Directory workingDirectory,
  Map<String, String> environment,
  List<String> arguments,
) async {
  final result = await Process.run(
    Platform.resolvedExecutable,
    arguments,
    workingDirectory: workingDirectory.path,
    environment: environment,
  );
  expect(
    result.exitCode,
    0,
    reason:
        'dart ${arguments.join(' ')} failed\n'
        'stdout:\n${result.stdout}\n'
        'stderr:\n${result.stderr}',
  );
  return result;
}

File: julia/test/unicode_rule_label_identity_routes_test.jl
# FUTURE-PARITY-BACKLOG.10.6.1.2 — exact downstream Julia label identity.

function _julia_unicode_identity_labels()
    labels = String[]
    for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["positive_fixtures"]
        push!(labels, String(fixture["label"]))
    end
    for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["distinct_fixtures"]
        for side in ("left", "right")
            label = String(fixture[side])
            if !(label in labels)
                push!(labels, label)
            end
        end
    end
    return labels
end

function _julia_unicode_identity_source(labels::AbstractVector{<:AbstractString})
    output = IOBuffer()
    for (index, label) in enumerate(labels)
        println(output, label, index == 1 ? "::" : ":")
        println(output, " /x/")
        println(output, " E { return(", JSON3.write(label), ") }")
        println(output)
    end
    return String(take!(output))
end

function _julia_unicode_identity_error(call)
    try
        call()
    catch error
        return error
    end
    return nothing
end

function _julia_unicode_identity_host_process(
    scratch,
    runner,
    generated,
    labels_path,
    private_depot,
)
    separator = Sys.iswindows() ? ';' : ':'
    parent_depot = get(ENV, "JULIA_DEPOT_PATH", "")
    depot_path = isempty(parent_depot) ?
        private_depot : string(private_depot, separator, parent_depot)
    load_path = join(
        [scratch, joinpath(REPO_ROOT, "julia"), "@stdlib"],
        separator,
    )
    command = `$(Base.julia_cmd()) --project=$scratch --startup-file=no --history-file=no --compiled-modules=no $runner $generated $labels_path`
    environment = copy(ENV)
    environment["JULIA_DEPOT_PATH"] = depot_path
    environment["JULIA_LOAD_PATH"] = load_path
    environment["JULIA_PKG_OFFLINE"] = "true"
    output = IOBuffer()
    errors = IOBuffer()
    process = run(pipeline(
        ignorestatus(setenv(command, environment));
        stdout = output,
        stderr = errors,
    ))
    return success(process), String(take!(output)), String(take!(errors))
end

const JULIA_UNICODE_IDENTITY_LABELS = _julia_unicode_identity_labels()
const JULIA_UNICODE_IDENTITY_SOURCE =
    _julia_unicode_identity_source(JULIA_UNICODE_IDENTITY_LABELS)

@testset "every positive and distinct label survives compiled artifacts" begin
    labels = JULIA_UNICODE_IDENTITY_LABELS
    @test length(JULIA_UNICODE_RULE_LABEL_CONTRACT["positive_fixtures"]) == 9
    @test length(JULIA_UNICODE_RULE_LABEL_CONTRACT["distinct_fixtures"]) == 2
    @test length(labels) == 10

    parsed = parse_spec(JULIA_UNICODE_IDENTITY_SOURCE)
    validate_spec(parsed)
    compiled = compile_spec(parsed)
    compiled_json = to_json(compiled)
    descriptor = to_descriptor_json(compiled)
    plan = build_generated_rule_plan(compiled)

    @test [rule.header.label for rule in parsed.rules] == labels
    @test compiled.definition_order == labels
    @test compiled.compiled_rule_order == labels
    @test Set(keys(compiled.rules_by_label)) == Set(labels)
    @test Set(keys(compiled_json["rules_by_label"])) == Set(labels)
    @test Set(keys(descriptor["spec"])) == Set(labels)
    @test descriptor["meta"]["definition_order"] == labels
    @test descriptor["meta"]["compiled_rule_order"] == labels
    @test [row.label for row in plan] == labels

    reconstructed_spec = from_json(
        SpecFile,
        JSON3.read(JSON3.write(to_json(parsed))),
    )
    reconstructed = compile_spec(reconstructed_spec)
    @test to_json(reconstructed) == compiled_json
    @test to_descriptor_json(reconstructed) == descriptor

    native_engine = LinkedSpecRuntimeEngine(reconstructed)
    @test runtime_parse(native_engine, "x").value == first(labels)
    for label in labels
        rule = compiled_rule(compiled, label)
        @test rule !== nothing
        @test rule.label == label
        @test runtime_parse(native_engine, "x"; top_rule = label).value == label
        @test execute_generated_parser_v2(
            compiled,
            plan,
            "x",
            "unicode-label/direct-plan.spec";
            top_rule = label,
        ) == label
    end

    for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["distinct_fixtures"]
        left = String(fixture["left"])
        right = String(fixture["right"])
        @test left != right
        @test compiled_rule(compiled, left) !== nothing
        @test compiled_rule(compiled, right) !== nothing
        @test compiled_rule(compiled, left) != compiled_rule(compiled, right)
    end
end

@testset "emitted source reconstructs and executes every exact label" begin
    labels = JULIA_UNICODE_IDENTITY_LABELS
    identity = "unicode-label/emitted-規則-𐐀.spec"
    compiled = compile_spec(parse_spec(JULIA_UNICODE_IDENTITY_SOURCE))
    emitted = emit_julia_source_v2(compiled, identity)
    encoded = match(r"const _COMPILED_SPEC_JSON_HEX = \"([0-9a-f]+)\"", emitted)
    @test encoded !== nothing
    payload = String(hex2bytes(encoded.captures[1]))
    reconstructed = compile_spec(from_json(SpecFile, JSON3.read(payload)))
    @test reconstructed.compiled_rule_order == labels
    @test [row.label for row in build_generated_rule_plan(reconstructed)] == labels

    mktempdir() do scratch
        private_depot = joinpath(scratch, "depot")
        mkpath(private_depot)
        write(
            joinpath(scratch, "Project.toml"),
            "name = \"JuliaUnicodeLabelEmittedHost\"\n" *
            "uuid = \"cdb67a3c-e949-4fa8-b661-a37ff4a7a224\"\n" *
            "version = \"0.1.0\"\n",
        )
        generated_path = joinpath(scratch, "generated_parser.jl")
        runner_path = joinpath(scratch, "runner.jl")
        labels_path = joinpath(scratch, "labels.json")
        write(generated_path, emitted)
        write(labels_path, JSON3.write(labels))
        write(
            runner_path,
            """
import JSON3
import LinkedSpecJulia

include(ARGS[1])
const Parser = LinkedSpecGeneratedParser
labels = JSON3.read(read(ARGS[2], String), Vector{String})
plan = Parser.plan()
Parser.validate_plan(plan)
values = [Parser.execute("x"; top_rule = label) for label in labels]
print(JSON3.write(Dict(
    "identity" => Parser.metadata().source_identity,
    "labels" => labels,
    "plan" => [row.label for row in plan],
    "values" => values,
)))
""",
        )
        passed, output, errors = _julia_unicode_identity_host_process(
            scratch,
            runner_path,
            generated_path,
            labels_path,
            private_depot,
        )
        @test passed
        @test isempty(errors)
        @test JSON3.read(output, Dict{String,Any}) == Dict{String,Any}(
            "identity" => identity,
            "labels" => labels,
            "plan" => labels,
            "values" => labels,
        )
    end
end

@testset "strict loading and primary commands preserve every exact label" begin
    labels = JULIA_UNICODE_IDENTITY_LABELS
    mktempdir() do scratch
        path = joinpath(scratch, "unicode-規則.spec")
        write(path, JULIA_UNICODE_IDENTITY_SOURCE)
        loaded = load_and_compile_spec(path_spec_request(path), SpecLoadOptions(scratch))
        @test loaded.loaded.source_text == JULIA_UNICODE_IDENTITY_SOURCE
        @test loaded.compiled.compiled_rule_order == labels
        @test !occursin(path, JSON3.write(to_json(loaded.compiled)))
        @test !occursin(path, JSON3.write(to_descriptor_json(loaded.compiled)))

        engine = create_engine(loaded)
        for label in labels
            @test runtime_parse(engine, "x"; top_rule = label).value == label
            output = IOBuffer()
            errors = IOBuffer()
            @test run_cli(
                [
                    "--inline-spec",
                    JULIA_UNICODE_IDENTITY_SOURCE,
                    "--input",
                    "x",
                    "--top-rule",
                    label,
                ];
                io = output,
                err = errors,
            ) == 0
            @test JSON3.read(chomp(String(take!(output))), String) == label
            @test isempty(String(take!(errors)))
        end

        supplementary = only(
            String(fixture["label"])
            for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["positive_fixtures"]
            if fixture["id"] == "supplementary"
        )
        output = IOBuffer()
        errors = IOBuffer()
        @test run_cli(
            [
                "--spec-file",
                path,
                "--input",
                "x",
                "--top-rule",
                supplementary,
            ];
            io = output,
            err = errors,
        ) == 0
        @test JSON3.read(chomp(String(take!(output))), String) == supplementary
        @test isempty(String(take!(errors)))
    end
end

@testset "selectors diagnostics and traces retain exact Unicode identity" begin
    selected = String(only(
        fixture["right"]
        for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["distinct_fixtures"]
        if fixture["id"] == "normalization_sensitive"
    ))
    missing = "規則Missing𐐀"
    identity = "unicode-label/diagnostic-規則.spec"
    compiled = compile_spec(parse_spec(JULIA_UNICODE_IDENTITY_SOURCE))
    engine = LinkedSpecRuntimeEngine(compiled)
    plan = build_generated_rule_plan(compiled)

    trace_output = IOBuffer()
    trace = LinkedSpecTraceEmitter(
        trace_config_enabled(LinkedSpecTraceDebug);
        stdout_io = trace_output,
    )
    @test runtime_parse(engine, "x"; top_rule = selected, trace = trace).value == selected
    @test any(
        event.topic == "julia_runtime:entry_rule_selection" &&
        occursin("requested=$selected", event.details) &&
        occursin("effective=$selected", event.details) &&
        occursin("basis=explicit_selector", event.details)
        for event in trace_events(trace)
    )
    native_trace_text = String(take!(trace_output))
    @test occursin("top_rule=$selected", native_trace_text)
    @test occursin(selected, native_trace_text)

    native_failure = _julia_unicode_identity_error() do
        runtime_parse(engine, "x"; top_rule = missing, trace = trace)
    end
    @test native_failure isa RuntimeInterpreterException
    if native_failure isa RuntimeInterpreterException
        @test to_json(native_failure.diagnostic) == Dict{String,Any}(
            "type" => "runtime_parser",
            "stage" => "select_entry_rule",
            "owner_stage" => "julia_runtime",
            "summary" => "Julia runtime entry-rule selection failed",
            "detail" => "entry rule '$missing' is not defined",
            "code" => "entry_rule_not_found",
            "top_rule" => missing,
            "entry_rule" => missing,
            "rule_label" => missing,
            "handler_source_label" => "julia_runtime:rule:$missing",
        )
    end
    @test any(
        event.topic == "julia_runtime:entry_rule_selection" &&
        occursin("requested=$missing", event.details)
        for event in trace_events(trace)
    )

    generated_failure = _julia_unicode_identity_error() do
        execute_generated_parser_v2(
            compiled,
            plan,
            "x",
            identity;
            top_rule = missing,
        )
    end
    @test generated_failure isa GeneratedSourceException
    if generated_failure isa GeneratedSourceException
        @test to_json(generated_failure) == Dict{String,Any}(
            "type" => "generated_source_error",
            "stage" => "select_entry_rule",
            "code" => "entry_rule_not_found",
            "summary" => "Generated Julia parser entry-rule selection failed",
            "source_identity" => identity,
            "entry_rule" => missing,
            "rule_label" => missing,
            "detail" => "entry rule '$missing' is not defined",
        )
    end

    mktempdir() do scratch
        trace_path = joinpath(scratch, "generated.trace")
        @test execute_generated_parser_with_trace_v2(
            compiled,
            plan,
            "x",
            LinkedSpecTraceConfig(
                level = LinkedSpecTraceDebug,
                trace_file = trace_path,
                sink_mode = LinkedSpecTraceRoute,
                reset_file = true,
            ),
            identity;
            top_rule = selected,
        ) == selected
        generated_trace = read(trace_path, String)
        @test occursin("requested=$selected", generated_trace)
        @test occursin("effective=$selected", generated_trace)
        @test occursin("top_rule=$selected", generated_trace)
        @test occursin(identity, generated_trace)
    end
end
